### PR TITLE
feat: replacement support for all managers

### PR DESCRIPTION
**Pull request by t-kulmburg**
As requested I am splitting my replacement pr #17476.
This pr contains the main replacement logic and lots of tests.

### CODE REVIEW

File: lib/workers/repository/update/branch/auto-replace.spec.ts [hunk 1]
```diff
@@ -25,12 +25,10 @@ describe('workers/repository/update/branch/auto-replace', () => {
     });
 
     beforeEach(() => {
-      upgrade = partial<BranchUpgradeConfig>({
-        // TODO: fix types (#7154)
-        ...(getConfig() as any),
-        manager: 'html',
-        packageFile: 'test',
-      });
+      // TODO: fix types (#7154)
```
[the remark this answers is not in the document]
Comment (rarkins) [reply lib/workers/repository/update/branch/auto-replace.spec.ts:28]: @Edouard-chin this is another instance of the bot making uninvited reviews. Unfortunately it's not possible to block a bot, and it was never installed so I cannot uninstall it, so I think I need you to take manual action

[the remark this answers is not in the document]
Comment (Edouard-chin) [reply lib/workers/repository/update/branch/auto-replace.spec.ts:28]: I uninstalled Catana on this repo. (You installed it a couple weeks ago , it wouldn't be able to receive webhooks from code push nor comment on this PR without a valid access token)

[the remark this answers is not in the document]
Comment (rarkins) [reply lib/workers/repository/update/branch/auto-replace.spec.ts:28]: Thanks. I think where I got confused is that it's `catana-dev` bot commenting. I went to github.com/apps/catana-dev and it was not installed. I guess it was github.com/apps/catana?

[the remark this answers is not in the document]
Comment (rarkins) [reply lib/workers/repository/update/branch/auto-replace.spec.ts:28]: Strangely when I went to github.com/apps/catana-dev it showed as uninstalled, like this:

image

BTW I assumed that apps could comment on PRs even without being installed, if they're public PRs (just like any user can)

[the remark this answers is not in the document]
Comment (Edouard-chin) [reply lib/workers/repository/update/branch/auto-replace.spec.ts:28]: I think if you visit https://github.com/apps/catana-dev GitHub would show whether you installed it on your personal (@rarkins account). You'd have had to visit https://github.com/renovatebot/renovate/settings/installations instead.

> BTW I assumed that apps could comment on PRs even without being installed [...]

Maybe, but the API call needs to be made while being authenticated with an installation access token (scoped or not to the organisation where the repo belongs, not sure).

**Comment by chezsmithy** [comment]
I'm so in need of replacement for docker. This feature is a slam dunk to help me solve for a number of deprecated docker images I need teams to swap out. Any ideas on ETA?

**Comment by t-kulmburg** [comment]
Any updates here? @JamieMagee

File: lib/config/utils.ts [hunk 2]
```diff
@@ -19,7 +19,8 @@ export function mergeChildConfig<
     if (
       option.mergeable &&
       childConfig[option.name] &&
-      parentConfig[option.name]
+      parentConfig[option.name] &&
+      parentConfig[option.name] !== childConfig[option.name]
```
Comment (viceice) [inline lib/config/utils.ts:23]: PLease do this in separate PR and add a testcase for it

**Review by viceice — CHANGES_REQUESTED** [review]

**Comment by JamieMagee** [comment]
> Any updates here? @JamieMagee

Apologies. I can find time to sit down and review both PRs in depth this weekend.

File: lib/config/utils.ts (hunk 2 above)
Comment (JamieMagee) [reply lib/config/utils.ts:23]: Is this still required after #18858?

File: lib/workers/repository/update/branch/auto-replace.ts [hunk 3]
```diff
@@ -147,9 +151,22 @@ export async function doAutoReplace(
   if (reuseExistingBranch) {
     return await checkExistingBranch(upgrade, existingContent);
   }
+  const replaceWithoutReplaceString = Boolean(
+    newName &&
+      newName !== depName &&
+      (!upgrade.replaceString ||
+        upgrade.replaceString?.indexOf(depName!) === -1)
```
Comment (JamieMagee) [inline lib/workers/repository/update/branch/auto-replace.ts:158]: This is a little dense, and hard to parse. Would you be open to using `@sindresorhus/is` to clarify the meaning a little? For example:

- `is.string(newName)` instead of `newName`
- `is.undefined(upgrade.replaceString)` instead of `!upgrade.replaceString`
- `upgrade.replaceString?.includes(depName)` instead of `upgrade.replaceString?.indexOf(depName!) === -1`

File: lib/workers/repository/update/branch/auto-replace.spec.ts [hunk 4]
```diff
@@ -181,5 +179,749 @@ describe('workers/repository/update/branch/auto-replace', () => {
...
+      upgrade.depName = 'github.com/hashicorp/example';
+      upgrade.newName = 'github.com/hashicorp/new-example';
+      upgrade.currentValue = 'v1.0.0';
+      upgrade.depIndex = 0;
+      upgrade.packageFile = 'modules.tf';
+      const res = await doAutoReplace(upgrade, hcl, reuseExistingBranch);
+      expect(res).toBe(hcl.replace(upgrade.depName, upgrade.newName));
+    });
+
+    it('works with old name in depname', async () => {
```
Comment (JamieMagee) [inline lib/workers/repository/update/branch/auto-replace.spec.ts:222]: Can you clarify what scenario this is covering?

**Review by JamieMagee — CHANGES_REQUESTED** [review]
Couple of small comments, but otherwise this looks good. Especially thanks for the extensive test coverage!

File: lib/config/utils.ts (hunk 2 above)
Comment (t-kulmburg) [reply lib/config/utils.ts]: no, it is not required anymore. i removed it

**Review by JamieMagee — DISMISSED** [review]

**Review by rarkins — COMMENTED** [review]
Any regression here would have a major impact. How confident is everybody and how widely has this been tested, not only on replacements but on existing autoreplace use cases?

**Review by rarkins — DISMISSED** [review]

File: lib/workers/repository/update/branch/auto-replace.spec.ts [hunk 5]
```diff
@@ -181,5 +179,749 @@ describe('workers/repository/update/branch/auto-replace', () => {
...
+    });
+
+    it('updates with ansible replacement', async () => {
+      const yml =
+        '- name: Container present\n' +
+        '  docker_container:\n' +
+        '    name: mycontainer\n' +
+        '    state: present\n' +
+        '    image: ubuntu:14.04\n' +
+        '    command: sleep infinity\n';
```
Comment (viceice) [inline lib/workers/repository/update/branch/auto-replace.spec.ts:250]: please use our new  `codeBlock` tagged template 

https://github.com/renovatebot/renovate/blob/848dc84244bbb41cedabd2329931a6d563fe6f97/lib/modules/manager/flux/extract.spec.ts#L102-L118

File: lib/workers/repository/update/branch/auto-replace.ts [hunk 6]
```diff
@@ -147,9 +152,22 @@ export async function doAutoReplace(
   if (reuseExistingBranch) {
     return await checkExistingBranch(upgrade, existingContent);
   }
+  const replaceWithoutReplaceString = Boolean(
+    is.string(newName) &&
+      newName !== depName &&
+      (is.undefined(upgrade.replaceString) ||
+        !upgrade.replaceString?.includes(depName!))
+  );
```
Comment (viceice) [inline lib/workers/repository/update/branch/auto-replace.ts:160]:
```suggestion
  const replaceWithoutReplaceString = (
    is.string(newName) &&
      newName !== depName &&
      (is.undefined(upgrade.replaceString) ||
        !upgrade.replaceString?.includes(depName!))
  );
```

remove explicit `Boolean` function call

**Review by viceice — CHANGES_REQUESTED** [review]

**Comment by HonkingGoose** [comment]
Our docs homepages has this list item:

> - Get replacement PRs to migrate from a deprecated dependency to the community suggested replacement (npm packages only)

Does merging this PR mean that you can now get replacements for old deprecated dependencies for all package managers? Of course the replacements still need to be programmed into the code for each package.

File: lib/workers/repository/update/branch/auto-replace.ts [hunk 7]
```diff
@@ -147,9 +152,21 @@ export async function doAutoReplace(
...
   }
+  const replaceWithoutReplaceString =
+    is.string(newName) &&
+    newName !== depName &&
+    (is.undefined(upgrade.replaceString) ||
+      !upgrade.replaceString?.includes(depName!));
   const replaceString = upgrade.replaceString ?? currentValue;
   logger.trace({ depName, replaceString }, 'autoReplace replaceString');
-  let searchIndex = existingContent.indexOf(replaceString!);
+  let searchIndex;
```
Comment (viceice) [inline lib/workers/repository/update/branch/auto-replace.ts:162]:
```suggestion
  let searchIndex: number;
```

**Comment by JamieMagee** [comment]
> Does merging this PR mean that you can now get replacements for old deprecated dependencies for all package managers? Of course the replacements still need to be programmed into the code for each package.

@HonkingGoose merging this PR will add replacement support for all managers _except_:

- bazel
- git-submodules
- gomod
- gradle
- hermit
- homebrew
- maven
- regex

File: lib/workers/repository/update/branch/auto-replace.spec.ts [hunk 8]
```diff
@@ -1,5 +1,6 @@
+import { codeBlock } from 'common-tags';
```
Comment (HonkingGoose) [inline lib/workers/repository/update/branch/auto-replace.spec.ts:1]: > > Does merging this PR mean that you can now get replacements for old deprecated dependencies for all package managers? Of course the replacements still need to be programmed into the code for each package.
> 
> @HonkingGoose merging this PR will add replacement support for all managers _except_:
> 
>     * bazel
>     * git-submodules
>     * gomod
>     * gradle
>     * hermit
>     * homebrew
>     * maven
>     * regex

Are these exceptions listed in the docs somewhere? I don't see any Markdown file changes in this PR. And I also don't see any docs tools code changes.

We should list those exceptions somewhere in our docs.

**Comment by shoddyknight** [comment]
Very keen on this, currently trying to migrate some internal gitlab-ci.yml, so want to replace

```yaml
includes:
- project: 'a/b/c'
    ref: 2-2-0
```

with either
```yaml
includes:
- project: 'z/x/y'
    ref: 2-2-0
```
or if needs be can tag new region
```yaml
includes:
- project: 'z/x/y'
    ref: 2-2-1
```

File: lib/workers/repository/update/branch/auto-replace.spec.ts (hunk 8 above)
Comment (t-kulmburg) [reply lib/workers/repository/update/branch/auto-replace.spec.ts:1]: The documentation currently says
```
This config option only works with the npm manager. We're working to support more managers, subscribe to issue [renovatebot/renovate#14149](https://github.com/renovatebot/renovate/issues/14149) to follow our progress.
```

How about changing this to 
```
This config option is currently not working with all managers. We're working to support more managers, subscribe to issue renovatebot/renovate#14149 to follow our progress.
```
and use #14149 to show whats currently working?


Or would you prefer a list directly inside the docs?

Comment (HonkingGoose) [reply lib/workers/repository/update/branch/auto-replace.spec.ts:1]: I prefer a list inside the docs. Can you also put a comment in the code so we know to update the list later?

Comment (t-kulmburg) [reply lib/workers/repository/update/branch/auto-replace.spec.ts:1]: Im not quite sure if thats what you meant but i added a comment at the beginning of the doAutoReplace function.

File: docs/usage/configuration-options.md [hunk 9]
```diff
@@ -1978,18 +1978,40 @@ Example setting source URL for package "dummy":
...
+Managers which currently do not support replacement:
+
+- bazel
+- git-submodules
+- gomod
+- gradle
+- hermit
+- homebrew
+- maven
+- regex
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1993]:
```suggestion
- `bazel`
- `git-submodules`
- `gomod`
- `gradle`
- `hermit`
- `homebrew`
- `maven`
- `regex`
```

File: docs/usage/configuration-options.md [hunk 10]
```diff
@@ -1978,18 +1978,40 @@ Example setting source URL for package "dummy":
 
 ### replacementName
 
-This config option only works with the `npm` manager.
+This config option is not working with all managers.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1981]:
```suggestion
This config option only works on some managers.
```

File: docs/usage/configuration-options.md [hunk 11]
```diff
@@ -1978,18 +1978,40 @@ Example setting source URL for package "dummy":
...
 We're working to support more managers, subscribe to issue [renovatebot/renovate#14149](https://github.com/renovatebot/renovate/issues/14149) to follow our progress.
 
+Managers which currently do not support replacement:
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1984]:
```suggestion
Managers which do not support replacement:
```

File: docs/usage/configuration-options.md [hunk 12]
```diff
@@ -1978,18 +1978,40 @@ Example setting source URL for package "dummy":
...
+- regex
+
 Use this field to define the name of a replacement package.
 Must be used with `replacementVersion` (see example below).
 You can suggest a new community package rule by editing [the `replacements.ts` file on the Renovate repository](https://github.com/renovatebot/renovate/blob/main/lib/config/presets/internal/replacements.ts) and opening a pull request.
 
 ### replacementVersion
 
-This config option only works with the `npm` manager.
+This config option is not working with all managers.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2001]:
```suggestion
This config option only works with some managers.
```

File: docs/usage/configuration-options.md [hunk 13]
```diff
@@ -1978,18 +1978,40 @@ Example setting source URL for package "dummy":
...
 We're working to support more managers, subscribe to issue [renovatebot/renovate#14149](https://github.com/renovatebot/renovate/issues/14149) to follow our progress.
 
+Managers which currently do not support replacement:
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2004]:
```suggestion
Managers which do not support replacement:
```

File: docs/usage/configuration-options.md [hunk 14]
```diff
@@ -1978,18 +1978,40 @@ Example setting source URL for package "dummy":
...
+
+- bazel
+- git-submodules
+- gomod
+- gradle
+- hermit
+- homebrew
+- maven
+- regex
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2013]:
```suggestion
- `bazel`
- `git-submodules`
- `gomod`
- `gradle`
- `hermit`
- `homebrew`
- `maven`
- `regex`
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: docs/usage/configuration-options.md [hunk 15]
```diff
@@ -1978,18 +1978,40 @@ Example setting source URL for package "dummy":
...
+- `bazel`
+- `git-submodules`
+- `gomod`
+- `gradle`
+- `hermit`
+- `homebrew`
+- `maven`
+- `regex`
+
 Use this field to define the name of a replacement package.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1995]:
```suggestion
Use the `replacementName` config option to set the name of a replacement package.
```

File: docs/usage/configuration-options.md [hunk 16]
```diff
@@ -1978,18 +1978,40 @@ Example setting source URL for package "dummy":
...
+- `bazel`
+- `git-submodules`
+- `gomod`
+- `gradle`
+- `hermit`
+- `homebrew`
+- `maven`
+- `regex`
+
 Use this field to define the version of a replacement package.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2015]:
```suggestion
Use the `replacementVersion` config option to set the version of a replacement package.
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
We can prevent the reader confusing the  `replacementName` and `replacementVersion` sections by saying which "field" they should use.

I'm also rewriting the sentence for extra clarity.

File: docs/usage/configuration-options.md (hunk 14 above)
Comment (rarkins) [inline docs/usage/configuration-options.md]: Can we avoid duplicating this list? Is it enough to list on `replacementName`?

Comment (chezsmithy) [reply docs/usage/configuration-options.md]: @HonkingGoose thoughts on the above question?

Comment (HonkingGoose) [reply docs/usage/configuration-options.md]: We should have a sentence like this in `replacementVersion` docs, and remove the list of managers from `replacementVersion`.

> For a list of managers which do not support replacement read the `replacementName` config option docs.

**Review by JamieMagee — APPROVED** [review]

**Comment by chezsmithy** [comment]
@t-kulmburg @rarkins looks like we might have all the approvals to merge? (I'm so very interested in this feature) 🥇

**Review by viceice — APPROVED** [review]

**Comment by t-kulmburg** [comment]
@rarkins let me know if there is anything more you want me to do here
